src/metabase/query_processor/middleware/fix_bad_references.clj

```diff
@@ -0,0 +1,86 @@
+  (:require [clojure.tools.logging :as log]
+            [clojure.walk :as walk]
+            [metabase.mbql.util :as mbql.u]
+            [metabase.query-processor.store :as qp.store]
+            [metabase.util :as u]
+            [metabase.util.i18n :refer [trs]]))
+
+(defn- find-source-table [{:keys [source-table source-query]}]
+  (or source-table
+      (when source-query
+        (recur source-query))))
+
+(defn- find-join-against-table [{:keys [joins source-query]} table-id]
+  (or (when source-query
+        (find-join-against-table source-query table-id))
+      (some (fn [join]
+              (when (= (find-source-table join) table-id)
+                join))
+            joins)))
+
+(defn- table [table-id]
+  (when table-id
+    (qp.store/fetch-and-store-tables! #{table-id})
+    (qp.store/table table-id)))
+
+(defn- fix-bad-references*
+  ([inner-query]
+   (fix-bad-references* inner-query inner-query (find-source-table inner-query)))
+
+  ([inner-query form source-table & sources]
+   (mbql.u/replace form
+     ;; don't replace anything inside source metadata.
+     (_ :guard (constantly ((set &parents) :source-metadata)))
+     &match
+
+     ;; if we have entered a join map and don't have `join-source` info yet, determine that and recurse.
+     (m :guard (every-pred map?
+                           :condition
+                           (fn [join]
+                             (let [join-source (find-source-table join)]
+                               (not (contains? (set sources) join-source))))))
+     (apply fix-bad-references* inner-query m source-table (cons (find-source-table m) sources))
+
+     ;; find Field ID fields whose Table IS NOT the source table (or not directly available in some `[:source-query+
+     ;; :source-table]` path that do not have `:join-alias` info
+     [:field
+      (id :guard (every-pred integer? (fn [id]
+                                        (let [{table-id :table_id} (qp.store/field id)]
+                                          (not (some (partial = table-id)
+                                                     (cons source-table sources)))))))
+      (opts :guard (complement :join-alias))]
+     (let [{table-id :table_id, :as field} (qp.store/field id)
+           {join-alias :alias}             (find-join-against-table inner-query table-id)]
+       (log/warn (u/colorize 'yellow (trs "Bad :field clause {0} for field {1} at {2}: clause should have a :join-alias. Guessing join {3}"
+                                          (pr-str &match)
+                                          (pr-str (format "%s.%s"
+                                                          (:name (table table-id))
+                                                          (:name field)))
+                                          (pr-str &parents)
+                                          (pr-str join-alias))))
+       (if join-alias
+         [:field id (assoc opts :join-alias join-alias)]
+         &match)))))
```

Comment:
my current reading of this is that if for some reason we don't find a join-alias, we just go back to the previous undefined behavior - is this correct? I can't find the exception-throwing anywhere and I can't trigger it with REPL shenanigans

Reply:
Sorry, you're right. This _did_ originally throw an Exception but in the end I removed that because it seemed better to let the query run and (possibly) work correctly than to have it fail completely. There are actually a bunch of malformed queries out there that we can't fix automatically that still work for some reason or another and breaking those seemed like a bad idea in the end.

I'll update the docstring